src/debug/nodejsDebugProvider.ts

```diff
@@ -0,0 +1,104 @@
+import * as path from "path";
+import * as vscode from "vscode";
+import { ChildProcess } from "child_process";
+
+import { IDebugProvider, PortInfo } from "./debugProvider";
+import * as config from '../components/config/config';
+import { Kubectl } from "../kubectl";
+import { IDockerfile } from "../docker/parser";
+import { kubeChannel } from "../kubeChannel";
+
+interface ExecResult {
+    readonly proxyProcess: ChildProcess;
+}
+const debuggerType = 'nodejs';
+
+export class NodejsDebugProvider implements IDebugProvider {
+    remoteRoot: String;
+    shell: string;
+
+    public getDebuggerType(): string {
+        return debuggerType;
+    }
+
+    public async isDebuggerInstalled(): Promise<boolean> {
+        return true;
+    }
+
+    public async startDebugging(workspaceFolder: string, sessionName: string, port: number): Promise<boolean> {
+
+        const debugConfiguration = {
+            type: "node",
+            request: "attach",
+            name: sessionName,
+            hostName: "localhost",
+            skipFiles: [
+                "<node_internals>/**/*.js"
+            ],
+            port
+        };
+        if (this.remoteRoot) {
+            debugConfiguration['remoteRoot'] = this.remoteRoot;
+        }
+        const currentFolder = vscode.workspace.workspaceFolders.find((folder) => folder.name === path.basename(workspaceFolder));
+        return await vscode.debug.startDebugging(currentFolder, debugConfiguration);
+    }
+
+    isSupportedImage(baseImage: string): boolean {
+        throw new Error("Method not implemented.");
+    }
+
+    resolvePortsFromFile(dockerfile: IDockerfile, env: {}): Promise<PortInfo> {
+        throw new Error("Method not implemented.");
+    }
+
+    public async resolvePortsFromContainer(kubectl: Kubectl, pod: string, podNamespace: string, container: string): Promise<PortInfo> {
+        this.shell = await this.suggestedShellForContainer(kubectl, pod, podNamespace, container);
+        await this.setNodeInspectMode(kubectl, pod, podNamespace, container);
+        this.remoteRoot = await this.tryGetRemoteRoot(kubectl, pod, podNamespace, container);
```

Comment:
You don't seem to use `remoteRoot` here - is this setting it up for use by some other method that gets called later?  I'd prefer to avoid setting required state in something that appears to be a simple getter, if we can.  (But if there is nowhere else we can do it then so be it.)

Reply:
`startDebugging` uses it. We could pass it back to the debugSession (like the port), but I think that since the `DebugProvider` instance lives only for the lifetime of the debug session, storing it in a member is fine. What do you think?

Reply:
I agree that storing it in a member is fine.  I just feel a bit awkward about modifying object state in a function called `resolvePortsFromContainer` (whose name implies that it is about reading the container settings to get a set of ports).  But if this is the place where we have the inputs we need (pod and container names) then I guess it is okay.  Perhaps in future we should amend the interface to provide a more natural home for this, but that can be out of scope for this PR.